proof/refine/ARM/Finalise_R.thy

```diff
@@ -2329,12 +2329,162 @@ lemma replyUnlink_invs'[wp]:
   unfolding invs'_def valid_state'_def
   by wpsimp
 
+crunches schedContextDonate
+  for replies_of'[wp]: "\<lambda>s. P (replies_of' s)"
+
+lemma isHeadSome:
+ "isHead (Some h) \<longleftrightarrow> (\<exists>p. h = Head p)"
+  by (clarsimp simp: isHead_def split: reply_next.split)
+
+lemma replyPrev_list_refs_of_replies':
+  "\<lbrakk>ko_at' reply replyPtr s; replyPrev reply = Some prevReply\<rbrakk>
+     \<Longrightarrow> (prevReply, ReplyPrev) \<in> list_refs_of_replies' s replyPtr"
+  by (clarsimp simp: obj_at'_def projectKOs list_refs_of_replies'_def opt_map_def
+                     list_refs_of_reply'_def
+              split: option.splits)
+
+lemma list_refs_of_replies'_eq:
+  "ko_at' reply replyPtr s \<Longrightarrow> list_refs_of_replies' s replyPtr = list_refs_of_reply' reply"
+  by (clarsimp simp: obj_at'_def projectKOs list_refs_of_replies'_def opt_map_def)
+
+lemma in_list_refs_of_replies'_E:
+  assumes in_list: "(p, tp) \<in> list_refs_of_replies' s replyPtr"
+  assumes nextR: "\<And>reply. tp = ReplyNext \<Longrightarrow> replyNext reply = Some (Next p)
+                            \<Longrightarrow> replies_of' s replyPtr = Some reply \<Longrightarrow> R"
+  assumes prevR: "\<And>reply. tp = ReplyPrev \<Longrightarrow> replyPrev reply = Some p
+                            \<Longrightarrow> replies_of' s replyPtr = Some reply \<Longrightarrow> R"
+  shows "R"
+  using in_list
+  apply (clarsimp simp: list_refs_of_replies'_def list_refs_of_reply'_def in_get_refs
+                 split: option.splits)
+  by (auto elim: nextR prevR simp: opt_map_def)
+
+lemma sym_refs_replyNext:
+  "\<lbrakk>sym_refs (list_refs_of_replies' s); replyPrev reply = Some prevReplyPtr;
+    ko_at' (replyTCB_update Map.empty reply) replyPtr s; ko_at' prevReply prevReplyPtr s\<rbrakk>
+     \<Longrightarrow> replyNext prevReply = Some (Next replyPtr)"
+  apply (drule sym_refs_replyNext_replyPrev_proj[where rp=prevReplyPtr and rp'=replyPtr])
+  by (clarsimp simp: opt_map_def obj_at'_def projectKOs split: option.splits)
```

Comment:
This also has extra things in the assumptions. The `replyTCB_update` part is irrelevant to the proof. And we don’t need to have `ko_at’`, either. The lemma is essentially a non-projection version of `sym_refs_replyNext_replyPrev_proj`, but we don’t need both versions. It should be possible to use the projection version instead of this one.

Reply:
Fixed, using non-projective version now. Removing this lemma from the PR.